packages/astro/src/core/app/pipeline.ts

```diff
@@ -41,4 +66,68 @@ export class AppPipeline extends Pipeline {
 	}
 
 	componentMetadata() {}
+	async getComponentByRoute(routeData: RouteData): Promise<ComponentInstance> {
+		const module = await this.getModuleForRoute(routeData);
+		return module.page();
+	}
+
+	async tryRewrite(payload: RewritePayload): Promise<[RouteData, ComponentInstance]> {
+		let foundRoute;
+
+		for (const route of this.#manifestData!.routes) {
+			if (payload instanceof URL) {
+				if (route.pattern.test(payload.pathname)) {
+					foundRoute = route;
+					break;
+				}
+			} else if (payload instanceof Request) {
+				const url = new URL(payload.url);
+				if (route.pattern.test(url.pathname)) {
+					foundRoute = route;
+					break;
+				}
+			} else {
+				if (route.pattern.test(decodeURI(payload))) {
+					foundRoute = route;
+					break;
+				}
+			}
+		}
+
+		if (foundRoute) {
+			const componentInstance = await this.getComponentByRoute(foundRoute);
+			return [foundRoute, componentInstance];
+		} else {
+			throw new Error('Route not found');
+		}
```

Comment:
```suggestion
		}
		throw new Error('Route not found');
```
nit: unneeded `else`